krabs/krabs/parser.hpp

```diff
@@ -142,7 +142,9 @@ namespace krabs {
         assert((pBufferIndex_ <= pEndBuffer_ && pBufferIndex_ >= schema_.record_.UserData) &&
                "invariant: we should've already thrown for falling off the edge");
 
-        assert((pBufferIndex_ == pEndBuffer_ ? lastPropertyIndex_ == totalPropCount
+        // accept that last property can be omitted from buffer. this happens if last property
+        // is string but empty and the provider strips the null terminator
+        assert((pBufferIndex_ == pEndBuffer_ ? ((totalPropCount - lastPropertyIndex_) <= 1)
```

Comment:
Interesting point. I admittedly have not yet come across events with two zero length strings yet but I think theoretically it might be possible. Would be totally okay with removing the assert completely. Should I do that then?

Reply:
Is there any indication in the schemas that certain fields may not appear in the record? That seems like a critical flaw if not. I would like to see if this behavior is documented somewhere.

Reply:
This one is based on the comments in tdh.h: https://github.com/thnetii/windows-sdk-headers/blob/master/um/tdh.h#L124 For this there was also already a test in krabs (parse_unicode_string_should_work_when_unicode_string_property_is_last_and_not_null_terminated) to cover this. But it did not fully encapsulate the issue as it only tries to parse the last value. And then the assert does not trigger as nothing from the message buffer has been consumed. The new test shows that behavior.

Reply:
There is an interesting code comment here: https://docs.microsoft.com/en-us/windows/win32/etw/using-tdhformatproperty-to-consume-event-data

`The length will be zero for variable length, null-terminated strings and structures.`

Reply:
@jrave thanks for pointing out the comment from `tdh.h`:

> Note that some event providers do not correctly nul-terminate the last string field in the event. While this is technically invalid, event decoders may silently tolerate such behavior instead of rejecting the event as invalid.

Given that the Microsoft comment specifically calls out the "last string field in the event" I am okay with your proposed change.